### ISSUE TITLE
Move to modern CMake target-based propagation of build information

### ISSUE THREAD

**Issue by bartlettroscoe** [TriBITSPub/TriBITS#299]
**Parent Issue:**
* #367

**Child Issues:**
* #433
* #443

**Blocked By:** trilinos/Trilinos#8498

## Description

One of the things holding back TriBITS and projects using TriBITS in accessing new features of CMake for which TriBITS still using variables to manage compiler options and include directories.  This issue is to move to using:

* target_compile_definitions()
* target_compile_features()
* target_compile_options()
* target_include_directories()
* target_link_directories()
* target_link_libraries()
* target_link_options()

and to move completely to targets instead of variables for each TriBITS (internal and external) package.

This needs to be done as part of a comprehensive refactoring as part of #63.

This refactoring should implement and respect [this proposed standard for `<Package>Config.cmake` files](https://docs.google.com/document/d/19gIZgPAfnK7RTdh0sMKr5Yeo1epPVKIfona-oxoCtsg).

Related to:
* SEPW-214

## Tasks

* [x] Add `PUBLIC` to `target_link_libraries()` in TriBITS intra-package linking
* [x] Add new library`${PACKAGE_NAME}::<lib-name>` targets and `${PACKAGE_NAME}::all_libs` INTERFACE library targets for TriBITS packages and `target_link_libraries( ... PUBLIC ... )` to define dependencies to upstream packages. (see PR #424)
* [x] Experiment with adding new `${TPL_NAME}::all_libs` INTERFACE GLOBAL library targets for TPLs and use `target_link_libraries( ... INTERFACE ... )` to define dependencies between TPLs.
* [x] Add include directories to all internal targets with `target_include_directories()`
* [x] Address problems with partial refactoring of TriBITS to modern CMake targets and get merge to Trilinos 'develop' to stick: #433 
* [x] Complete refactoring of TriBITS targets to use modern CMake targets for all dependencies and managing include directories (no more using variables `<Package>_INCLUDE_DIRS`, `<Package>_TPL_INCLUDE_DIRS`, `<Package>_LIBRARIES` or `<Package>_LIBRARY_DIRS`).
* [x] Strip out all old machinery in TriBITS involving vars `<Package>_INCLUDE_DIRS`, `<Package>_TPL_INCLUDE_DIRS`, `<Package>_LIBRARIES` and `<Package>_LIBRARY_DIRS`.
* [x] Do some testing with Trilinos, Albany, Nalu, and SPARC.
* [x] Create and post new Trilinos PR to snaphsot updated TriBITS 'master' branch: https://github.com/trilinos/Trilinos/pull/10614
* [ ] Get PR https://github.com/trilinos/Trilinos/pull/10614 to merge (see **Task** there)

**Comment by bartlettroscoe** [comment]
This refactoring will make it easier to integrate Kokkos and KokkosKernels into Trilinos as described in the below email from @jjwilke.

----

From: Wilke, Jeremiah J <jjwilke@sandia.gov> 
Sent: Tuesday, February 4, 2020 4:18 PM
To: Wolf, Michael M <mmwolf@sandia.gov>; Rajamanickam, Sivasankaran <srajama@sandia.gov>; Bartlett, Roscoe A <rabartl@sandia.gov>; Willenbring, James M <jmwille@sandia.gov>; Trott, Christian Robert <crtrott@sandia.gov>; Teranishi, Keita <knteran@sandia.gov>
Subject: To-do list for Trilinos

Below would be my (very specific) action items to get where we need to be.

•	Move to CMake 3.13 (allows link_options and better with flag deduplication)
•	Change variables to lists instead of strings (where appropriate)
•	Change include_directories/compile_options to target_include_directories(….)
•	Remove all dependencies between packages that rely on variables. Only use target_link_libraries(…) to propagate flags

I would guess this could be done in a month? Once this is done, it should be straightforward to design a system where Kokkos could be a TPL.

-Jeremy

**Comment by bartlettroscoe** [comment]
The Trilinos issue to allow the upgrade of the minimum version of CMake from 3.10 to 3.13 for TriBITS as well is trilinos/Trilinos#6752.

**Comment by bartlettroscoe** [comment]
@jjwilke, @sethrj,

This brings up another issue.  These refactorings of TriBITS to use target-centric features in CMake may make it very difficult to maintain backward comparability with-respect to the build system.  For example, it may not be possible keep supporting the generation of `Makefile.export.<Package>` files for customers like INL MOOSE.  Trilinos backward-comparability requirements might make that difficult to allow.

**Comment by jjwilke** [comment]
I think we can probably make this work? Everything that needs to be in the Makefile will end up in `INTERFACE_COMPILE_OPTIONS`, `INTERFACE_LINK_LIBRARIES`, etc. We could probably make a function 
````
FUNCTION(EXPORT_TARGET target)
...
````
That doesn't rely on variables like `Kokkos_CXX_FLAGS` and instead calls `get_property` to figure everything out.

**Comment by bradking** [comment]
This will also be relevant to https://github.com/TriBITSPub/TriBITS/issues/127#issuecomment-549912354.

**Comment by sethrj** [comment]
FWIW, SCALE/Exnihilo is only using the exported tribits config.cmake files for the downstream [enrico project](https://github.com/enrico-dev/enrico), not any makefile.export.

I'm definitely OK with a transition plan for removing backward compatibility gradually 😬

**Comment by jjwilke** [comment]
I will say (and maybe @bradking already knows how to do this) - the reason I stopped short of exporting a Makefile.kokkos was generator expressions.

**Comment by bradking** [comment]
> stopped short of exporting a Makefile.kokkos was generator expressions in the property lists.

Modern CMake target-based usage requirements mean that much of the logic to convert project CMake code into the final build system command-lines is handled by CMake's generators.  The input to that process is not exportable to `Makefile.*`, and the output is only available in generated build systems inside CMake-generated build trees.  Easy export to Makefile snippets was not a design goal.

**Comment by bartlettroscoe** [comment]
> Easy export to Makefile snippets was not a design goal.

It occurred to me that it still might be possible to produce `Makefile.*` files. What you do is you create a dummy CMake project that pulls in `TrilinosConfig.cmake` or `<Package>Config.cmake` and then you use your own compiler and liner wrappers to capture the compiler and linker arguments and use that to generate the `Makefile.export.*` files.  That is almost exactly what the ASC Sierra project does with their "bake" tool to generate Makefiles and build.ninja files given output from a Bjam build system.  Crude but that would work.

**Comment by bartlettroscoe** [comment]
FYI: I just got feedback from lead INL MOOSE framework lead that they dropped testing against Trilinos (because CASL has ended) and don't plan on supporting Trilinos going forward (unless something changes).  Therefore, because I think we could add support for `Makefile.export.*` files like a described [above](https://github.com/TriBITSPub/TriBITS/issues/299#issuecomment-582585296), and it is not needed now, I will remove testing and support for this in TriBITS when I do this refactoring.  Yea!

**Comment by bartlettroscoe** [comment]
FYI: It just occurred to me that the SNL Sierra project might be using Makefile.export.Trilinos for its custom bJam/bake build system that uses the native Trilinos TriBITS/CMake build system.  I have reached out to them to see if that is the case.

**Comment by bartlettroscoe** [comment]
FYI: Got back a response from the SNL Sierra DevOps lead that Sierra actually does not use the Makefile.export.Trilinos file.   The actually manually maintain a set of BJam files for building Trilinos and they get the info they need from that.

Therefore, it looks like we don't need to keep generating Makefile.export.Trilinos files based on SNL Sierra needs.  That is good news :-)

**Comment by bartlettroscoe** [comment]
Part of this refactoring is in the PR #300 from @bradking.  It removes explicit handing of `-std` compiler options.  It sets the global var `CMAKE_CXX_STANDARD` and it sets:

```
TARGET_COMPILE_FEATURES(<libtarget> PUBLIC <cxx_std>)
```

to that matching standard.

**Comment by bartlettroscoe** [comment]
Someone just showed me the option [CMAKE_EXPORT_COMPILE_COMMANDS](https://cmake.org/cmake/help/v3.13/variable/CMAKE_EXPORT_COMPILE_COMMANDS.html?highlight=cmake_export_compile_commands) which generated a `comple_commands.json` file after the generate phase.  From that you can easily extract the information needed for `Makefile.export.*` files (but that would have to be created by a build target, after configure and generate were complete).  Now to get other things out of the way so I can get into this.

**Comment by bartlettroscoe** [comment]
FYI: I just realized that PETSc is still using the `Makefile.export.Trilinos` file to get the list and order of Trilinos libraries.  See:

* https://gitlab.com/petsc/petsc/-/blob/7ea10ee11db2e07e7f75d402c9c0eb0db5382105/config/BuildSystem/config/packages/Trilinos.py#L288

So I think unless PETSc switches over to CMake, we are going to need to re-implement support for at least the `Makefile.export.<Project>` file as part of this refactoring :-(

**Comment by sethrj** [comment]
Or just require PETSc to use older trilinos/tribits version until they switch?

**Comment by bartlettroscoe** [comment]
> Or just require PETSc to use older trilinos/tribits version until they switch?

Not sure that is a viable option because of the ECP xSDK and E4S that includes PETSc and Trilinos.  But to get by, someone can just manually generate a `Makefile.export.Trilinos` file in the Spack Trilinos `package.py` file and that would satisfy PETSc for now for Spack and ECP.

And I have no idea what the timetable would be for PETSc to switch to CMake (or even if that is a firm plan).

**Comment by bartlettroscoe** [comment]
FYI: Looks like the refactored TriBITS may need to need to keep defining the vars `TPL_<TPLName>_INCLUDE_DIRS` due to code like:

* https://github.com/trilinos/Trilinos/pull/7368/files#diff-f71bd309778da4697b8611f1a6f94839R22

I wonder how much code like this exists in Trilinos and other projects using Trilinos?

**Comment by mperego** [comment]
@bartgol This would affect Albany and the export of the library for MPAS, right?

**Comment by bartlettroscoe** [comment]
FYI: Xyce should be getting some help from Kitware to transition fully to CMake so hopefully the export Makefile requirement for TriBITS and Trilinos can go away completely.

**Comment by bartlettroscoe** [comment]
FYI: trilinos/Trilinos#8498

**Comment by bartlettroscoe** [comment]
I guess I am technically starting on this Issue today.  I am upgrading TriBITS to require CMake 3.17+ given that Trilinos has already upgraded is min version to 3.17 (see trilinos/Trilinos#8660).  This will involve cleaning up a good bit of code and getting rid of various workarounds and hacks for older versions of CMake.

**Comment by bartlettroscoe** [comment]
@keitat, do you think that you can have some of the subcontractor staff looking into Trilinos xSDK issues help review PRs related to this and TriBITS changes in general?  I desperately need someone to help review TriBIITS changes that really knows CMake well.

**Comment by keitat** [comment]
Stay tuned.  I will let you know the github username of the people from NGA.

**Comment by bartlettroscoe** [comment]
FYI: It looks like there is an important internal customer that would be significantly impacted if Trilinos looses support for Makefile.export.Trilinos in the next few months (or longer).  This refactoirng can't wait any longer so if support for Makefile.export.Trilinos has to be maintained, then we might just have to bit the bullet and implement an approach like they did for ParaView/VTK shown in:

* https://gitlab.kitware.com/paraview/paraview/-/blob/master/Clients/CommandLineExecutables/paraview-config.in

which is outlined in https://github.com/trilinos/Trilinos/issues/8498#issuecomment-825349056 which is basically what I suggested [above](https://github.com/TriBITSPub/TriBITS/issues/299#issuecomment-582585296) last year.  The fact that Kitware has already implemented this approach for a major project gives me some comfort that this may not be as big of a hack as what I was thinking this was.

But note that this file Makefile.export.Trilinos would get generated at build time, not configure time, and use cases like the integration of INL MOOSE into VERA and demonstrated with the [WrapExternal](https://github.com/TriBITSPub/TriBITS/tree/master/tribits/examples/TribitsExampleProject/packages/wrap_external) package would not longer work (or would have to work very differently with the configure of the wrapped external project done at build time, not configure time of the outer TriBITS CMake project).

Just goes to show that once an important piece of software acquires a feature that is important to some customer, that you can never remove support for that feature :-(

**Comment by mathstuf** [comment]
> The fact that Kitware has already implemented this approach for a major project gives me some comfort that this may not be as big of a hack as what I was thinking this was.

I wouldn't call it a "big hack", but it is a hack in the sense that it doesn't *solve* the problem. For example, VTK's object factory stuff is absent there, so that is still manual. There are probably other usage requirements that are missing. If a project has Fortran-specific usage requirements, the synthesized project has to use Fortran for those to be exposed.

Basically, the example project needs to encompass the use cases that exist then have ways of teasing out what is wanted. That's where the "hackiness" comes in: you're making tiny examples then trying to extrapolate it for full usage which doesn't always work. I would stick huge disclaimers on this that CMake is the preferred way to use Trilinos and that this exporting mechanism is a best-effort attempt at bridging the gap.

**Comment by bartlettroscoe** [comment]
FYI: That internal SNL customer still using `Makefile.export.*` files for a daily integration with Trilinos 'master' is going to try switching that build over to using their new CMake build system for their application code.   They will try to get that done in the next few weeks.  Then they will plan to officially switch to CMake for all of their builds before their next major upgrade of Trilinos based on the current Trilinos 'master' branch.

Therefore, I need to wait until they give me the green light before I merge a version of TriBITS to Trilinos 'develop' that drops support for `Makefile.export.*` files.

**Comment by bartlettroscoe** [comment]
FYI: After talking with that internal customer still using `Makefile.export.*` we may be able to switch them over to use CMake for their daily integration builds against Trilinos 'develop' and avoid needing to support `Makefile.export.*`, even in the short-term.  For details, see https://github.com/trilinos/Trilinos/issues/8498#issuecomment-834681399.  Fingers crossed that transition to CMake for the customer will be completed soon so we are not blocked from dumping support for `Makefile.export.*` and get this TriBITS refactoring to use modern CMake done.  But we can't push an updated version of TriBITS to Trilinos that drops support for `Makefile.export.*` to Trilinos 'develop' until that customer's Xyce 'master' + Trilinos 'develop' integration build is switched over to use CMake.  I just don't want to be doing this work on long-lived a branch of TriBITS like I did with the branch [63-merge-tpls-packages](https://github.com/TriBITSPub/TriBITS/tree/63-merge-tpls-packages) that still is not merged in.

**Comment by bartlettroscoe** [comment]
Good news.  As stated in https://github.com/trilinos/Trilinos/issues/8498#issuecomment-850503201, it looks like we can finally drop support for `Makefile.export.*` files from Trilinos (and therefore TriBITS)!   I promised to keep support in the Trilinos 'develop' through 5/10/2021 but that is just a little under 2 weeks from now.

I realized that I can do a lot of the refactoring to create the new targets supporting modern CMake incrementally and safety without breaking backward compatibility initially to see how this works before actually needing to drop support for `Makefile.export.*`.   For example, I can create targets `<PackageName>::all_libs` for all of the TPLs and packages and ensure these targets actually work to link executables in a dummy TriBITS-lite example/test project.

So given where I am with the recent refactorings/cleanups in #63 as part of epic #367, I am ready to start extending TriBITS to provide fully modern CMake targets as per[ this proposed standard](https://docs.google.com/document/d/19gIZgPAfnK7RTdh0sMKr5Yeo1epPVKIfona-oxoCtsg/edit#heading=h.6v29dm1zgr8x) next week!

**Comment by bartlettroscoe** [comment]
FYI: I hit another snag in this refactoring work.  It seems that there may be a deficiency in CMake (even CMake 3.21.3) that does not propagate `INTERFACE_INCLUDE_DIRECTORIES` in the IMPORTED library targets generated by the `export(EXPORT ...)` command needed for the `<Package>Config.cmake` files in the build tree (needed for some existing use cases).  To try to get some help from Kitware, I created:

* https://gitlab.kitware.com/cmake/cmake/-/issues/22751 (So everyone can see this including other CMake users)
* https://gitlab.kitware.com/snl/project-1/-/issues/189 (To get this on the docket for our SNL Kitware contract work)

NOTE: I can likely workaround this by manually looping over the library targets for a package and copying over the `INTERFACE_INCLUDE_DIRECTORIES` target property.  But I would really do not want to do that since that just keeps up complexity in TriBITS that I am trying to get rid of with this refactoring.

**Comment by bartlettroscoe** [comment]
FYI: TriBITS PR #424 converts internal TriBITS targets to use modern CMake and the updated internal targets and generated `<Package>Config.cmake` files mostly satisfies the proposed convention in #342.  As noted in PR #424, I still need to address the handling of external packages/TPLs before I can rip out all of the old complex TriBITS logic and complete this issue #299 but I have a pretty good plan for that and my recent experiences gives me hope that will go okay.

Now if I could just find someone to review the documentation and examples in PR #424 and get some early feedback.

**Comment by bartlettroscoe** [comment]
FYI: Related to #340, as described in #427, there are issues with TriBITS-generated `<tplName>Config.cmake` files when `FindTPL<tplName>Config.cmake` modules call `find_package(<tplName>)`.  One issue raised in #427 is that `find_package()` will look under `CMAKE_INSTALL_PREFIX`.  That is a tricky issue that I will need to address somehow.

**Comment by bartlettroscoe** [comment]
CC: @mperrinel, @marcinwrobel1986, @keitat 

While debugging the errors with the updated TriBITS with Trilinos with Albany I came across a bit of an issue.  The problem is that it seems that application codes like Albany have not been always following the suggested usage of pulling in libraries using the variables `<PackageName>_LIBRARIES` but instead have been, in some cases, directly linking to the non-namespaced IMPORTED targets for libraries that are defined by the found `<Package>Config.cmake` files (included by the `TrilinosConfig.cmake` file through `find_package(Trilinos)`).  For example, Albany/src/CMakeLists.txts had code like:

https://github.com/sandialabs/Albany/blob/4e50f9bb54c88d1f589be232f363599a8dfd578c/src/CMakeLists.txt#L196-L203

showing:

```
add_executable(xml2yaml utility/xml2yaml.cpp)
...
target_link_libraries(xml2yaml teuchosparameterlist)
...
target_include_directories(xml2yaml SYSTEM PUBLIC
                          "${Trilinos_INCLUDE_DIRS};${Trilinos_TPL_INCLUDE_DIRS}")
```

When the old TriBITS implementation, `teuchosparameterlist` is an actual IMPORTED target and CMake uses this to create a successful link line for the `xml2yaml` executable in the Albany build system.

But with the new TriBITS, there is no non-namespaced IMPORTED target called `teuchosparameterlist` (it is the namespaced target `TeuchosParameterList::teuchosparameterlist`).   Therefore, when CMake processes the line

```
target_link_libraries(xml2yaml teuchosparameterlist)
```

with the updated `<Package>Config.cmake` files for the updated TriBITS, there is no existing target of that name.  And in this case, CMake assumes that `teuchosparameterlist` is just some dumb library out on the system and it puts a `-lteuchosparameterlist` on the link line as demonstrated in the detailed debugging in **Detailed debugging notes** in:

* https://cee-gitlab.sandia.gov/rabartl/tribitsrefactoring/-/issues/1#note_2118925

This is a very "helpful" feature of CMake's `target_link_libraries()` command as documented under the section **A plain library name** at:

* https://cmake.org/cmake/help/v3.17/command/target_link_libraries.html#overview

The fix for this is simple.  Applications should just have used the proper TriBITG-defined variables `${<Package>_LIBRARIES}` and linked against that instead as shown in the PR:

* https://github.com/sandialabs/Albany/pull/778

in lines:

https://github.com/sandialabs/Albany/blob/a4ddf552088a876f908b0721010d38feca0d6a4a/src/CMakeLists.txt#L200-L207

showing:

```
add_executable(xml2yaml utility/xml2yaml.cpp)
...
target_link_libraries(xml2yaml PUBLIC ${TeuchosParameterList_LIBRARIES})
...
target_include_directories(yaml2xml SYSTEM PUBLIC
  "${TeuchosParameterList_INCLUDE_DIRS};${TeuchosParameterList_TPL_INCLUDE_DIRS}")
```

And that works with both the old and the new TriBITS.

The variables:

* `<Package>_LIBRARIES`
* `<Package>_INCLUDE_DIRS`
* `<Package>_TPL_INCLUDE_DIRS`

are the TriBITS interface that I have been targeting maintaining (and the PR https://github.com/sandialabs/Albany/pull/778 and the testing associated with that demonstrates that still works).

I think the problem here is that the APP teams did not know how to find Trilinos using `find_package(Trilinos)` and then just link against a subset of the packages that are found.  Therefore, they just guessed, and in the case of Albany, grabbed the raw targets instead of linking against the proper varaibles.

The issue is that when Trilinos support for `TrilinosConfig.cmake` and the `<Package>Config.cmake` files was first created (by developers other than me), on one documented how they were supposed to be used or how customer APPs/packages were supposed to link against just a subset of package libraries.  I did not get involved in writing this documentation until just last year and by that point the ship had sailed for all of these APPs.

So that this point, we have one of two options to move forward from here:

* **Option-1:** Try to add back the old non-namespaced IMPORTED library targets like `teuchosparameterlist` in addition to the new namespaced targets `TeuchosParamterList::teuchosparameterlist` (and of course the `TeuchosParamterList::all_libs` target which customers are supposed to be linking against).

* **Option-2:** Send out emails informing the application teams of incorrect usage of TriBITS and ask them to check their CMakeLists.txt files for direct linking against raw imported targets and fix them to be `${<Package>_LIBRARIES}`.

If we go with ***option-1***, then that should provide better (hopefully about perfect) backward compatibility.  But then how do we later remove these non-namespaced targets and help the application codes to move to the new proper namespaced targets?

If we go with ***option-2***, then the APPs will experience nasty build errors like was seen for Albany as demonstrated in:

* https://cee-gitlab.sandia.gov/rabartl/tribitsrefactoring/-/issues/1#note_2118055

and it is not clear what is happening.

Note that once the updated TriBITS is merged to Trilinos, then APPs can just link against `<PackageName>::all_libs` and can stop using the include dirs altogether.  That would make this Albany/src/CMakeLists.txt code just be:

```
add_executable(xml2yaml utility/xml2yaml.cpp)
...
target_link_libraries(xml2yaml PUBLIC TeuchosParameterList::all_libs)
```

To decide between ***option-1*** and ***option-2***, I need to get advice from some experienced CMake people to get their opinions on how to proceed.  I will ask the NextGen people and my Kitware contractors as well.

In the meantime, I will see if this issue might be impacting Nalu as well (which is another APP with a public Git repo).

**Comment by bartlettroscoe** [comment]
Okay, I just found that there is `DEPRECATION` target property for CMake 3.17+:

* https://cmake.org/cmake/help/v3.22/prop_tgt/DEPRECATION.html

With that, we could go with ***option-1*** [above](https://github.com/TriBITSPub/TriBITS/issues/299#issuecomment-1011558738) and deprecate these old non-namespaced CMake targets so that, at configure time, if someone tried to link against a deprecated IMPORTED target, then it would generate a warning.

Given this discovery, I think that is the way to go with ***option-1*** and set the `DEPRECATION` target property.  These `<libname>` targets would be ALIAS targets against the proper namespaced targets `<Package>::<libname>`.  The question is, how do we generate those targets in the `<Package>Config.cmake` files?

**Comment by mathstuf** [comment]
> But then how do we later remove these non-namespaced targets and help the application codes to move to the new proper namespaced targets?

Make it depend on the requested version of Trilinos (or the package in question). If that is at least the version which provides the new targets, do not create the backwards compat bits. If it is explicitly lower, consider dropping the `DEPRECATION` property (because that target needs to be used to work across the requested version range). After "enough time" (i.e., when "everyone" could have been expected to migrate), this deprecation can become unconditional. If it is left empty, always add the `DEPRECATION` property.

> Given this discovery, I think that is the way to go with option-1 and set the DEPRECATION target property. These <libname> targets would be ALIAS targets against the proper namespaced targets <Package>::<libname>.

This sounds reasonable.

> The question is, how do we generate those targets in the <Package>Config.cmake files?

There's the list of libraries that goes into `all_libs` already? Can't that be reused?

**Comment by bradking** [comment]
I agree with ***option-1*** for the reasons posted above.  However, ***option-1*** and ***option-2*** are not mutually exclusive.  Even if Trilinos were not changing to the namespaced targets, the application teams should still replace their direct non-namespaced target usage with `${<Package>_LIBRARIES}` to avoid using undocumented names.  The variable references will then work with old versions of Trilinos too.  A deprecation warning from ***option-1*** will help teams implement ***option-2*** if they don't want to raise their minimum required version of Trilinos.

**Comment by bartlettroscoe** [comment]
@mathstuf,

> There's the list of libraries that goes into `all_libs` already? Can't that be reused?

Yes, I know how to get that list but I was hopping that I could just get CMake to generate these targets automatically with the export() command.  But given how the targets are defined and used currently using the namespacing, I am not sure If CMake will let me do this.  I will experiment with this and see what CMake will allow me to do.  If nothing else, I think I know how to create these ALIAS DEPRECATION non-namespaced targets directly in the generated `<Package>Config.cmake` file.  (But again, I would rather if CMake would just do this for me with the right arguments to the right commands.)

@bradking,

>  Even if Trilinos were not changing to the namespaced targets, the application teams should still replace their direct non-namespaced target usage with `${<Package>_LIBRARIES}` to avoid using undocumented names.  The variable references will then work with old versions of Trilinos too.

Right, but the question is if we should expect non-compliant customers to refactor their CMakeLists.txt files twice.  Once to refactor from the current non-namespaced targets:

```
target_link_libraries(xml2yaml teuchosparameterlist)
```

to the variables  `${<Package>_LIBRARIES}` :

```
target_link_libraries(xml2yaml PUBLIC ${TeuchosParameterList_LIBRARIES})
```

and then a second time to remove usage of variables and go with modern namespaced CMake targets  `<Package>::all_libs`  by changing to:

```
target_link_libraries(xml2yaml PUBLIC TeuchosParameterList::all_libs)
```

(or `Trilinos::TeuchosParamterList_libs` as per [DRAFT: Proposed Standard for PackageConfig.cmake files for inter-package Interoperability](https://docs.google.com/document/d/19gIZgPAfnK7RTdh0sMKr5Yeo1epPVKIfona-oxoCtsg)).

**Comment by bartlettroscoe** [comment]
@bradking and @mathstuf,

Now I understand this move of modern CMake to namespaced target names.  From "Professional CMake: 10th edition"

> Another important aspect of names having a double-colon (::) is that CMake will always treat them
as the name of an alias or imported target. Any attempt to use such a name for a different target
type will result in an error. Perhaps more usefully though, when the target name is used as part of
a target_link_library() call, if CMake doesn’t know of a target by that name, it will issue an error at
generation time. Compare this to an ordinary name which CMake will treat as a library assumed to
be provided by the system if it doesn’t know of a target by that name. This can lead to the error only
becoming apparent much later at build time.

That is what we are suffering from here.  Old CMake used non-namespaced targets and if a target `<some-lib>` become undefined, `target_link_libraries( ... <some-lib>)` will just assume that `<some-lib>` is out on the system and put in a `-l<some-lib>` on the link line.  That makes it impossible to catch errors in target names at configure or generation time.  You only find out about errors at link time where it is very difficult to debug what is happening (unless you know to watch for this).

Given this information, I think that you should **never** link against a non-namespaced target in a call to `target_link_libraries()`.  CMake just provides not checking whatsoever at configure or generation time with raw non-namespaced names.  Someone should have a list of basic CMake dos and don'ts and this needs to be on that list (both as "do create and link against namespaced targets" and "don't link against non-namespaced targets").

**Comment by bradking** [comment]
> you should never link against a non-namespaced target in a call to `target_link_libraries()`

FYI, CMake 3.23 adds an optional [LINK_LIBRARIES_ONLY_TARGETS](https://cmake.org/cmake/help/git-master/prop_tgt/LINK_LIBRARIES_ONLY_TARGETS.html) check to verify that plain names passed to `target_link_libraries` are actually targets.

**Comment by bradking** [comment]
> create these ALIAS DEPRECATION non-namespaced targets directly in the generated <Package>Config.cmake file.

FYI, CMake 3.18 is required to create an ALIAS target that references a non-global imported target.  ALIAS targets were not originally meant for this use case.

An alternative is to create a non-namespaced `IMPORTED` `INTERFACE` library that just sets `INTERFACE_LINK_LIBRARIES` to the namespaced target's name.

**Comment by bradking** [comment]
> hoping that I could just get CMake to generate these targets automatically...I am not sure If CMake will let me do this.

I don't think it will.  While one might be able to call `install(EXPORT)` twice, once with a namespace and once without, doing so would create two independent definitions of each target that don't de-duplicate and could double-up linking.

I think generating the compatibility targets yourself in `TribitsProjectConfigTemplate.cmake.in` or similar may work best.

**Comment by bartlettroscoe** [comment]
> FYI, CMake 3.23 adds an optional LINK_LIBRARIES_ONLY_TARGETS check to verify that plain names passed to target_link_libraries are actually targets.

That is great news!  Unfortunately, we can't turn this on for Trilinos right now due to people taking advantage of raw library names (see https://github.com/trilinos/Trilinos/pull/9978#issuecomment-982166363).

@bradking, would you recommend people change from passing raw `<libname>` to `target_link_libraries()` such as for:

```
 -D TPL_BLAS_LIBRARIES:PATH="${OPENBLAS_PATH}/lib/libopenblas.a;gfortran;gomp;m"
```

to passing `-l<libname>`:

```
 -D TPL_BLAS_LIBRARIES:PATH="${OPENBLAS_PATH}/lib/libopenblas.a;-lgfortran;-lgomp;-lm"
```

?

Is `-l<libname>` portable on Windows?

(I did not even realize that `target_link_libraries()` accepted raw lib names like this until https://github.com/trilinos/Trilinos/pull/9978#issuecomment-982166363 was posted.  I had never used that in 13+ years of my CMake usage.  And yes, I know people should be using `find_packge()` modules but that is a hard sell when people already have working configure scripts and people have really struggled with some of the `Find<Package>.cmake` modules over the years.)

One of the issues we are struggling with here is that early versions of CMake were incredibly permissive so attempts we make to improve things and move to modern CMake often result in errors that are not caught until compile or link time where they are very hard for some people to debug.  But I guess everyone is in the same boat.  (But Trilinos has a lot of customers that configure, build, and then link against Trilinos on all types of crazy systems by thousands of non-CMake users across the planet.)

**Comment by bartlettroscoe** [comment]
> I think generating the compatibility targets yourself in `TribitsProjectConfigTemplate.cmake.in` or similar may work best.

@bradking. I can do that, but that means that I will need to add some machinery to keep a list of targets for which to create these `INTERFACE IMPORTED` targets when generating the `<Package>Config.cmake` file from TribitsPackageConfigTemplate.cmake.in`.  But I am already gathering that list of targets to form the `<Package>::all_libs` target at:

https://github.com/TriBITSPub/TriBITS/blob/c8abb338d90d9f4065cd415eca01c7212a754084/tribits/core/package_arch/TribitsPackageMacros.cmake#L718-L729

I just need to pass that list `packageLibsInAllLibsList` to the generated file `<Package>Config.cmake` and let it loop over to form the non-namespaced `INTERFACE IMORTED` targets.

(NOTE: All of this code will get cleaned up significantly once we are done with this transition to full modern CMake inside of TriBITS.  Getting rid of all of the variables and global properties will be one of the major goals.)

**Comment by bradking** [comment]
Raw library names are actually the preferred way to refer to toolchain-provided libraries, rather than `find_library`, so that the toolchain can handle selecting the proper library file from its builtin paths.  `LINK_LIBRARIES_ONLY_TARGETS` documents a way to handle such libraries via `IMPORTED_LIBNAME` on an interface library, but only when the project code knows the names itself.

> Is `-l<libname>` portable on Windows?

No.  That would just pass through directly to the link line as a flag, and the MSVC linker does not understand such syntax.

Who is responsible for passing `${TPL_BLAS_LIBRARIES}` to a `target_link_libraries` call?  Is project code free to do that, or does TriBITS handle it somewhere?  In the latter case, it might be able to transform the content.

**Comment by bartlettroscoe** [comment]
> Who is responsible for passing `${TPL_BLAS_LIBRARIES}` to a `target_link_libraries` call?  Is project code free to do that, or does TriBITS handle it somewhere?  In the latter case, it might be able to transform the content.

@bradking, that is a good point.  The refactoring that I just finished actually does process the list in `${TPL_BLAS_LIBRARIES}` to create `IMPORTED` library targets and does not rely on passing raw library names to `target_link_libraries()`.  So with the upgraded TriBITS, in theory, Trilinos could set `LINK_LIBRARIES_ONLY_TARGETS=TRUE`.  I will create a backlog item for that.

**Comment by bartgol** [comment]
I am not reading the whole conversation, but I read that some changes won't happen due to downstream apps relying on "old" cmake patterns. FWIW, I wanted to make it clear that on the Albany side we are ok if Trilinos moves to moder cmake and breaks our builds. We will fix them up. Having more robust (and self-explanatory) cmake logic is definitely healthy, and I'm willing to take a few hours to adapt to whatever change you guys might do.

I know we're not the only/main trilinos customer, but I figured stating it clearly could help (especially if other apps are on the same page).

**Comment by bartlettroscoe** [comment]
Consensus at meeting today: go with ***option-1*** and add back non-namespaced targets and add a good `DEPRECATION` message that mentions the usage of `${<Package>_LIBRARIES}` to build against old and new TriBITS/Trilinos or just use the updated TriBITS/Trilinos with usage `<Package>::all_libs` and drop usage of the `<Package>_INCLUDE_DIRS` and `<Package>_TPL_INCLUDE_DIRS` variables.

**Comment by bartlettroscoe** [comment]
>  FWIW, I wanted to make it clear that on the Albany side we are ok if Trilinos moves to modern cmake and breaks our builds. We will fix them up. Having more robust (and self-explanatory) cmake logic is definitely healthy, and I'm willing to take a few hours to adapt to whatever change you guys might do.

@bartgol, it is good that some customers are willing to absorb non-backward compatible changes.  Long term, this has to happen if we are going to move to more robust modern CMake as a community.  This issue is how painful does this need to be for everyone involved? For the case of linking against the old non-namespaced targets, that resulted in a link-time error what was not trivial to debug (at least not for me).  We want to provide a smoother path to get customers to upgrade their usage of Trilinos for modern CMake and encourage them to adopt modern CMake.  If it was a clear configure-time error, then I think that is likely okay.  But a obscure link time error is not okay (at least not in my opinion).

**Comment by bartlettroscoe** [comment]
CC: @mperrinel, @marcinwrobel1986, @keitat 

FYI: A build problem with SPARC on the Vanguard system 'stsria' for the partial refactoring of TriBITS to use modern CMake targets in handling `INTERACE_INCLUDE_DIRECTORIES` has exposed a serious break in backwards compatibility in moving to modern CMake in how all IMPORTED targets include directories are handled as `SYSTEM` directories and included on the compile line with `-isystem` (which changes the search order for header files and how warnings are expressed from headers under these directories).  The problem and some potential (partial) workarounds are described in the new issue  #443.

From my study and analysis, it seems to be impossible to maintain backward compatibility with this TriBITS refactoring to modern CMake.  It seems that CMake would need to be extended in order to return backwards compatibility.  I need to consult with the Kitware CMake experts on how the advice we address this.

**Comment by bartlettroscoe** [comment]
FYI: Given that the lingering TriBITS snapshot into Trilinos 'develop' in Trilinos PR #9978 breaks backward compatibility (see #443), we might as well just wait to merge #9978 and charge ahead with completing the transition of TriBITS to modern CMake targets.

**Comment by bartlettroscoe** [comment]
FYI: With the merge of #479, this is essentially done. All of the TriBITS tests pass and I tested this against Albany and Nalu (see [internal tribitsrefactoring#4](https://cee-gitlab.sandia.gov/rabartl/tribitsrefactoring/-/issues/4) and [internal tribitsrefactoring#5](https://cee-gitlab.sandia.gov/rabartl/tribitsrefactoring/-/issues/5)).  I will test against SPARC next and will create a new TriBITS PR if I find any other issues.

But I will put this in review.

NOTE: I will not be able to take his out of review until TriBITS 'master' is synced to Trilinos 'develop' and we get feedback from customers.

**Comment by bartlettroscoe** [comment]
FYI: I created PR trilinos/Trilinos#10614 to update the snapshot of TriBITS 'master' into Trilinos 'develop' that switches to modern CMake targets.

**Comment by bartlettroscoe** [comment]
NOTE: PR trilinos/Trilinos#10614 was just merged.  I will leave this issue "In Review" for a few weeks to collect any lingering issues that may not have been caught yet.  I will work very hard to help resolve issues ASAP so we don't back out upgraded TriBITS.

**Comment by bartlettroscoe** [comment]
With the merge of trilinos/Trilinos#10614 two days ago, it exposed that there were still downstream CMake projects at SNL that were generating export makefiles for Trilinos, but indirectly by looping over `${Trilinos_LIBRRAIES}` (which is now just a list of namespaced targets).

**Comment by bartlettroscoe** [comment]
Below is a response to an internal Trilinos customer about supporting export makefiles.  This approach really has nothing to do with TriBITS.

----

If you really want to support export makefiles, an approach that will work is outlined in:

   https://github.com/TriBITSPub/TriBITS/issues/299#issuecomment-582585296
   https://github.com/TriBITSPub/TriBITS/issues/299#issuecomment-828678848
   https://github.com/TriBITSPub/TriBITS/issues/299#issuecomment-828692051

If supporting makefiles is critical, this is something you should take to the Trilinos Leaders and get a contactor to do.  The ability to generate an export Makefile from `find_package(<Package>)` is really independent from TriBITS and could work for nearly any CMake-generated `<Package>Config.cmake` file.  It is just that Kitware does not want to support such functionality.

**Comment by mathstuf** [comment]
> could work for nearly any CMake-generated <Package>Config.cmake file. It is just that Kitware does not want to support such functionality.

I've done it, but it is specific to each project. One for ParaView can be seen here: https://gitlab.kitware.com/paraview/paraview/-/blob/master/Clients/CommandLineExecutables/paraview-config.in

**Comment by bartlettroscoe** [comment]
FYI: There are a few issues with the merge of trilinos/Trilinos#10614 that are covered in trilinos/Trilinos#10774.  Once those issues are resolved, I think it is safe to close this issue.

**Comment by bartlettroscoe** [comment]
With the merge of Trilinos PR #trilinos/Trilinos#10930, I believe the transition to modern CMake for TriBITS and Trilinos is complete!

Closing as complete!
